### PR TITLE
Allow parallel execution: put execution to artifacts

### PR DESCRIPTION
**Pull request by mnecas**
This is another solution for parallel execution others (#50, #48).
Difference compared to #50 that instead of creating `private_data_dir` is created in artifacts instead of tmp folder. 

WDYT @jmolmo @pcuzner?
cc @mwperina @dangel101

### CODE REVIEW

File: runner_service/services/playbook.py [hunk 1]
```diff
@@ -214,8 +215,17 @@ def start_playbook(playbook_name, vars=None, filter=None, tags=None):
 
     # this should just be run_async, using 'run' hangs the root logger output
     # even when backgrounded
+
+    artifacts_dir = os.path.join(configuration.settings.playbooks_root_dir, "artifacts")
+    if configuration.settings.max_artifacts is not None:
+        remove_oldest_artifacts(artifacts_dir)
```
Comment (mwperina) [inline runner_service/services/playbook.py:221]: I think we should define a standalone cron like job to remove old artifacts and not removing them on each playbook execution

File: runner_service/services/playbook.py [hunk 2]
```diff
@@ -166,7 +167,7 @@ def cb_event_handler(event_data):
         # role is not a fixed attribute
         role_value = ''
         if 'role' in event_data:
-            role_value =  runner_cache[ident]['role'] = event_data['event_data'].get('role', '') # noqa
+            role_value = runner_cache[ident]['role'] = event_data['event_data'].get('role', '')  # noqa
```
Comment (mwperina) [inline runner_service/services/playbook.py]: How is this indentation change relevant?

File: runner_service/services/playbook.py [hunk 3]
```diff
@@ -182,24 +183,24 @@ def cb_event_handler(event_data):
                 else:
                     # we have a valid failure to report
                     event_metadata = event_data['event_data']
-                    runner_cache[ident]['failures'][event_metadata.get('host')] = event_data # noqa
+                    runner_cache[ident]['failures'][event_metadata.get('host')] = event_data  # noqa
```
Comment (mwperina) [inline runner_service/services/playbook.py]: How is this indentation change relevant?

File: runner_service/services/playbook.py [hunk 4]
```diff
@@ -182,24 +183,24 @@ def cb_event_handler(event_data):
...
 
     # populate the event cache
     if 'runner_ident' in event_data and \
-       'uuid' in event_data and ident in event_cache:
+            'uuid' in event_data and ident in event_cache:
```
Comment (mwperina) [inline runner_service/services/playbook.py]: How is this indentation change relevant?

File: runner_service/services/playbook.py (hunk 1 above)
Comment (mnecas) [reply runner_service/services/playbook.py:221]: When using cron we would not have constant number of artifacts and the effect would be same,
Unless we would want to define the maximum age of the artifact, but that would require larger changes.

Comment (mwperina) [reply runner_service/services/playbook.py:221]: Yes, I think we should introduce configuration option in config.yaml to define maximum age of artifacts (by default for example 30 days) and customers with high load could change the default value as needed.

So having a job which is started during ansible-runner-service startup can perform cleanup and then schedule it's own next execution within maximum age interval is the solution I'd prefer.

Comment (mnecas) [reply runner_service/services/playbook.py:221]: Okay, this will be a larger change. I will revert changes here and create separate PR and continue there.

**Comment by mnecas** [comment]
1. I thought that the artifacts//env is what we agreed on, what do you think should be better layout?
Was not able to reproduce the artifacts/artifacts//* in dev env.

2. Fixed the inventory.

**Comment by pcuzner** [comment]
@mnecas when we spoke about putting env under the artifacts I was meaning artifacts/<play-uuid>/env so within the run's output, you have the env that was used - so it's all togetther. I thought that's what we discussed - apologies for any confusion.

**Comment by pcuzner** [comment]
Yep - can see the inventory now, but hitting the problem I thought we would hit with regard to the ssh keys. the env dir is where runner picks the keys up from but they're not there - when I try to run using a sudo account
fatal: [rhcs4-installer]: UNREACHABLE! => {"changed": false, "msg": "Data could not be sent to remote host \"rhcs4-installer\". Make sure this host can be reached over ssh: Permission denied (publickey,gssapi-keyex,gssapi-with-mic,password).\r\n", "unreachable": true}
fatal: [rhcs4-aio]: UNREACHABLE! => {"changed": false, "msg": "Data could not be sent to remote host \"rhcs4-aio\". Make sure this host can be reached over ssh: Permission denied (publickey,gssapi-keyex,gssapi-with-mic,password).\r\n", "unreachable": true}

the ssh keys are in the system placed in /usr/share/ansible-runner-service/env. My assumption is that with this change, since the env dir is now somewhere else, these keys are missing..and we don't want to copy them inside every env dir either...does this make sense to you? symlink?
 
I'd like to ensure that resolving the mutliple execution problem, still supports non-root users and different keys. This was a base requirement for RHCS with ceph-ansible.

@jmolmo can you test this PR out too, just in case I'm doing something crazy!

**Comment by mnecas** [comment]
Found the issue [1] will probably need to remove the if statement so it every time passes the ssh from cmdline. I think putting it to cmdline would be better than symlinks, a user could still check the keys in artifacts//env/cmdline where would be a path to the keys.

I did not have this issue because I had a specific folder with ssh keys.

[1] https://github.com/ansible/ansible-runner-service/blob/master/runner_service/services/playbook.py#L255-L258

**Comment by jmolmo** [comment]
> Yep - can see the inventory now, but hitting the problem I thought we would hit with regard to the ssh keys. the env dir is where runner picks the keys up from but they're not there - when I try to run using a sudo account
> fatal: [rhcs4-installer]: UNREACHABLE! => {"changed": false, "msg": "Data could not be sent to remote host "rhcs4-installer". Make sure this host can be reached over ssh: Permission denied (publickey,gssapi-keyex,gssapi-with-mic,password).\r\n", "unreachable": true}
> fatal: [rhcs4-aio]: UNREACHABLE! => {"changed": false, "msg": "Data could not be sent to remote host "rhcs4-aio". Make sure this host can be reached over ssh: Permission denied (publickey,gssapi-keyex,gssapi-with-mic,password).\r\n", "unreachable": true}
> 
> the ssh keys are in the system placed in /usr/share/ansible-runner-service/env. My assumption is that with this change, since the env dir is now somewhere else, these keys are missing..and we don't want to copy them inside every env dir either...does this make sense to you? symlink?
> 
> I'd like to ensure that resolving the mutliple execution problem, still supports non-root users and different keys. This was a base requirement for RHCS with ceph-ansible.
> 
> @jmolmo can you test this PR out too, just in case I'm doing something crazy!

@pcuzner i cant find the problem ... I have tested  the ARS nginx container without any problem ... so this LGTM. Can you share the kind of tests you're doing ?

**Comment by jmolmo** [comment]
LGTM

**Comment by mwperina** [comment]
Resolves https://github.com/ansible/ansible-runner-service/issues/47
